src/copy.rs

```diff
@@ -149,42 +150,68 @@ where
 {
     let (mut rd, mut wd) = downstream.split_into_buffered_reader();
     let (mut ru, mut wu) = upstream.split_into_buffered_reader();
-
     let downstream_to_upstream = async {
-        let res = copy_buf(&mut rd, &mut wu, stats, false).await;
+        let report = |e: io::Error| {
```

Comment:
report as a name was pretty confusing to me. It holds a fn which converts io::ErrorKind into proxy::Error, yeah? I think something like `into_proxy_error` or `into_senderror` would clarify.